src/System.IO.FileSystem/src/System/IO/PathHelpers.Windows.cs

```diff
@@ -2,80 +2,45 @@
-using System.Diagnostics;
-
 namespace System.IO
 {
-    internal static partial class PathHelpers
+    internal static partial class PathInternal
     {
-        // Trim trailing whitespace, tabs etc but don't be aggressive in removing everything that has UnicodeCategory of trailing space.
-        // string.WhitespaceChars will trim more aggressively than what the underlying FS does (for ex, NTFS, FAT).
-        internal static readonly char[] TrimEndChars = { (char)0x9, (char)0xA, (char)0xB, (char)0xC, (char)0xD, (char)0x20, (char)0x85, (char)0xA0 };
-        internal static readonly char[] TrimStartChars = { ' ' };
-
-        internal static bool ShouldReviseDirectoryPathToCurrent(string path)
-        {
-            // In situations where this method is invoked, "<DriveLetter>:" should be special-cased
-            // to instead go to the current directory.
-            return path.Length == 2 && path[1] == ':';
-        }
-
-        // ".." can only be used if it is specified as a part of a valid File/Directory name. We disallow
-        //  the user being able to use it to move up directories. Here are some examples eg
-        //    Valid: a..b  abc..d
-        //    Invalid: ..ab   ab..  ..   abc..d\abc..
-        //
-        internal static void CheckSearchPattern(string searchPattern)
+        internal static unsafe int GetRootLength(ReadOnlySpan<char> path)
         {
-            for (int index = 0; (index = searchPattern.IndexOf("..", index, StringComparison.Ordinal)) != -1; index += 2)
+            fixed (char* p = &path.DangerousGetPinnableReference())
             {
-                // Terminal ".." or "..\". File and directory names cannot end in "..".
-                if (index + 2 == searchPattern.Length ||
-                    PathInternal.IsDirectorySeparator(searchPattern[index + 2]))
-                {
-                    throw new ArgumentException(SR.Arg_InvalidSearchPattern, nameof(searchPattern));
-                }
+                return (int)GetRootLength(p, (uint)path.Length);
             }
         }
+    }
 
-        internal static string NormalizeSearchPattern(string searchPattern)
+    internal static partial class PathHelpers
+    {
+        internal static bool ShouldReviseDirectoryPathToCurrent(string path)
         {
-            Debug.Assert(searchPattern != null);
-
-            // Win32 normalization trims only U+0020.
-            string tempSearchPattern = searchPattern.TrimEnd(PathHelpers.TrimEndChars);
-
-            // Make this corner case more useful, like dir
-            if (tempSearchPattern.Equals("."))
-            {
-                tempSearchPattern = "*";
-            }
-
-            CheckSearchPattern(tempSearchPattern);
-            return tempSearchPattern;
+            // In situations where this method is invoked, "<DriveLetter>:" should be special-cased
+            // to instead go to the current directory.
+            return path != null && path.Length == 2 && path[1] == ':';
```

Comment:
Do we need to check what path[0] is?

Reply:
This check I'm not thrilled about in general. Not sure what the original intent was. Adding the check wouldn't do much other than make `!:` not "work" when it does now. I'll open an issue to track this weird behavior.